db/memtable.cc

```diff
@@ -583,54 +584,64 @@ bool MemTable::InsertKey(std::unique_ptr<MemTableRep>& table, KeyHandle handle,
+  // entry after [value, +inf)
+  bool done = false;
   for (; iter->Valid(); iter->Next()) {
     const Slice& ikey = iter->key();
     Slice tombstone_start_key = ExtractUserKey(ikey);
-    Slice tombstone_end_key = iter->value();
-    if (!last_start_key.empty() &&
-        user_comparator->Compare(last_start_key, tombstone_start_key) == 0) {
+    if (!last_key.empty() &&
+        user_comparator->Compare(tombstone_start_key, last_key) <= 0) {
       continue;
     }
-    if (!iter->IsKeyPinned()) {
-      pinned_slices_.emplace_back(tombstone_start_key.data(),
-                                  tombstone_start_key.size());
-      tombstone_start_key = pinned_slices_.back();
-    }
-    if (!iter->IsValuePinned()) {
-      pinned_slices_.emplace_back(tombstone_end_key.data(),
-                                  tombstone_end_key.size());
-      tombstone_end_key = pinned_slices_.back();
-    }
+    Slice tombstone_end_key = iter->value();
     if (user_comparator->Compare(tombstone_start_key, value) >= 0) {
       break;
     }
-    last_start_key = tombstone_start_key;
     std::vector<SequenceNumber> tombstone_seqs;
     FragmentedRangeTombstoneList::DecodeRangeTombstoneValue(&tombstone_end_key,
                                                             &tombstone_seqs);
+    last_key = tombstone_end_key;
     int start_compare = user_comparator->Compare(key, tombstone_start_key);
     if (start_compare > 0) {
       int end_compare = user_comparator->Compare(value, tombstone_end_key);
       if (end_compare < 0) {
+        to_insert[std::make_pair(tombstone_start_key, key)].insert(
+            tombstone_seqs.begin(), tombstone_seqs.end());
         auto k = std::make_pair(key, value);
         to_insert[k].insert(tombstone_seqs.begin(), tombstone_seqs.end());
         to_insert[k].insert(s);
         to_insert[std::make_pair(value, tombstone_end_key)].insert(
             tombstone_seqs.begin(), tombstone_seqs.end());
+        done = true;
+        break;
       } else if (end_compare == 0) {
+        to_insert[std::make_pair(tombstone_start_key, key)].insert(
+            tombstone_seqs.begin(), tombstone_seqs.end());
         auto k = std::make_pair(key, value);
         to_insert[k].insert(tombstone_seqs.begin(), tombstone_seqs.end());
         to_insert[k].insert(s);
+        done = true;
+        break;
       } else if (user_comparator->Compare(key, tombstone_end_key) < 0) {
+        to_insert[std::make_pair(tombstone_start_key, key)].insert(
+            tombstone_seqs.begin(), tombstone_seqs.end());
         auto k = std::make_pair(key, tombstone_end_key);
         to_insert[k].insert(tombstone_seqs.begin(), tombstone_seqs.end());
         to_insert[k].insert(s);
-        to_insert[std::make_pair(tombstone_end_key, value)].insert(s);
+        key = tombstone_end_key;
       } else {
-        to_insert[std::make_pair(key, value)].insert(s);
+        continue;
```

Comment:
If we reach here it means `tombstone_end_key <= key`. I think it can only happen if the we are at the first iteration of the loop and there is no range tombstones overlapping with the one being inserted (so SeekForPrev gives a range tombstone with `tombstone_end_key <= key`. If that is the case, then we can safely `break` here?

Reply:
yes